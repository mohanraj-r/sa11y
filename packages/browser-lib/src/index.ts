/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { base, recommended, full } from '@sa11y/preset-rules';
export { exceptionListFilter } from '@sa11y/format';
export { assertAccessible } from '@sa11y/assert';

export const nameSpace = 'sa11y';
