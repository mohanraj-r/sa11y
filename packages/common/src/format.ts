/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { AxeResults } from './axe';

export const errMsgHeader = 'accessibility issues found';

/**
 * Filter to post-process a11y results from axe
 */
export interface Filter {
    (violations: AxeResults, ...args: never[]): AxeResults;
}

// TODO (refactor): constrain rule id to known rule ids e.g using string literal, keyof, in etc
//  e.g. https://stackoverflow.com/a/54061487
// const ruleIDs = getRules().map((ruleObj) => ruleObj.ruleId);
// type RuleID = keyof ruleIDs;
// type RuleID = typeof ruleIDs[number];
type RuleID = string;
type CssSelectors = string[];
/**
 * Exception list of map of rule to corresponding css targets that needs to be filtered from a11y results.
 */
export type ExceptionList = Record<RuleID, CssSelectors>;
