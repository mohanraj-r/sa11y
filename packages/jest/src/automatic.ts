/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { AxeResults, log } from '@sa11y/common';
import { getViolationsJSDOM } from '@sa11y/assert';
import { A11yError } from '@sa11y/format';

/**
 * Options for Automatic checks to be passed to {@link registerSa11yAutomaticChecks}
 */
export type AutoCheckOpts = {
    runAfterEach?: boolean;
    cleanupAfterEach?: boolean;
    consolidateResults?: boolean;
    // TODO (feat): add support for optional exclusion of selected tests
    // excludeTests?: string[];
};

/**
 * Default options when {@link registerSa11yAutomaticChecks} is invoked
 */
const defaultAutoCheckOpts: AutoCheckOpts = {
    runAfterEach: true,
    cleanupAfterEach: true,
    consolidateResults: true,
};

/**
 * Run accessibility check on each element node in the body using {@link toBeAccessible}
 * @param opts - Options for automatic checks {@link AutoCheckOpts}
 */
export async function automaticCheck(opts: AutoCheckOpts = defaultAutoCheckOpts): Promise<void> {
    const violations: AxeResults = [];
    // Create a DOM walker filtering only elements (skipping text, comment nodes etc)
    const walker = document.createTreeWalker(document.body, NodeFilter.SHOW_ELEMENT);
    let currNode = walker.firstChild();
    // Using fake timer results in async timeout issue with axe
    jest.useRealTimers();
    try {
        while (currNode !== null) {
            // TODO (spike): Use a logger lib with log levels selectable at runtime
            // console.log(
            //     `â™¿ [DEBUG] Automatically checking a11y of ${currNode.nodeName}
            //      for test "${expect.getState().currentTestName}"
            //      : ${expect.getState().testPath}`
            // );
            violations.push(...(await getViolationsJSDOM(currNode)));
            currNode = walker.nextSibling();
        }
    } finally {
        if (opts.cleanupAfterEach) document.body.innerHTML = ''; // remove non-element nodes
        // TODO (spike): Disable stack trace for automatic checks.
        //  Will this affect all errors globally?
        // Error.stackTraceLimit = 0;
        A11yError.checkAndThrow(violations, { deduplicate: opts.consolidateResults });
    }
}

/**
 * Register accessibility checks to be run automatically after each test
 * @param opts - Options for automatic checks {@link AutoCheckOpts}
 */
export function registerSa11yAutomaticChecks(opts: AutoCheckOpts = defaultAutoCheckOpts): void {
    if (opts.runAfterEach) {
        log('Registering sa11y checks to be run automatically after each test');
        afterEach(async () => {
            await automaticCheck(opts);
        });
    }
}
