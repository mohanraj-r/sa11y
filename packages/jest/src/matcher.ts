/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { matcherHint, printReceived } from 'jest-matcher-utils';
import { adaptA11yConfig } from './setup';
import { A11yCheckableContext, assertAccessible } from '@sa11y/assert';
import { A11yError, Options } from '@sa11y/format';
import { A11yConfig, recommended } from '@sa11y/preset-rules';

// Type def for custom jest a11y matchers
// Ref: https://jestjs.io/docs/en/expect.html#expectextendmatchers
declare global {
    // eslint-disable-next-line @typescript-eslint/no-namespace
    namespace jest {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        interface Matchers<R> {
            toBeAccessible(config?: A11yConfig): Promise<CustomMatcherResult>;
        }
    }
}

const matcherHintMsg = `expected document to have no accessibility violations but found`;
const expectedMsg = `0 issues`;
const formatOptions: Options = {
    a11yViolationIndicator: 'â­•',
    helpUrlIndicator: 'ðŸ”—',
    highlighter: printReceived,
};

/**
 * Jest expect matcher to check DOM for accessibility issues
 * @param received - DOM or HTML Element to be tested for accessibility. Defaults to current DOM.
 * @param config - A11yConfig to be used to test for accessibility. Defaults to recommended.
 */
export async function toBeAccessible(
    received: A11yCheckableContext = document,
    config: A11yConfig = recommended
): Promise<jest.CustomMatcherResult> {
    let isAccessible = true;
    // a11yError can be 'undefined' when used with `not` matcher and there is no error
    let a11yError: A11yError = new A11yError([]);
    let receivedMsg = expectedMsg;

    // TODO (Improvement): Can we detect if this is invoked async and error if not ?
    try {
        await assertAccessible(received, adaptA11yConfig(config));
    } catch (e) {
        if (e instanceof A11yError) {
            a11yError = e;
            isAccessible = false;
            receivedMsg = `${a11yError.length} issues`;
        } else {
            throw e;
        }
    }

    return {
        pass: isAccessible,
        message: (): string =>
            matcherHint(
                `${toBeAccessible.name}: ${matcherHintMsg} ${receivedMsg}: \n\n ${a11yError.format(formatOptions)}`,
                `${a11yError.length} issues`,
                expectedMsg
            ),
    };
}
