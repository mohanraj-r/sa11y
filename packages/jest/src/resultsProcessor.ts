/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { AggregatedResult, AssertionResult, SerializableError } from '@jest/test-result/build/types';
import { TestResult } from '@jest/reporters';
import { A11yError } from '@sa11y/format';
import { buildFailureTestResult } from '@jest/test-result';

type FailureDetail = {
    error?: A11yError;
};

// Map of test suite name to test results
const consolidatedErrors = new Map<string, AssertionResult[]>();

// suite.name = [Sa11y WCAG-`2.0`-Level-`A` SC-`1.1.1` `image-alt`]: `testSuite.testFilePath`
// test.name = CSS Selectors[0]
// test.failure.message =
//  first line: Accessibility issue found:`description`
//  rest: Help URL, CSS Selectors, List of tests `test.name`, Sa11y Help URL

function addTestResult(testResult: AssertionResult, testSuite: TestResult) {
    // const sa11yResults = [] as TestResult[];
    // Copy test suite meta-data and clear test results
    // const sa11yTestSuite = { ...testSuite, testResults: [] };
    const suiteName = testSuite.testFilePath.substring(testSuite.testFilePath.lastIndexOf('/') + 1);
    testResult.failureDetails.forEach((failure) => {
        let error = (failure as FailureDetail).error;
        // If using circus test runner https://github.com/facebook/jest/issues/11405#issuecomment-843549606
        if (error === undefined) error = failure as A11yError;
        if (error.name === A11yError.name) {
            // TODO : What happens if ever there are multiple failureDetails? Are there ever?
            error.violations.forEach((violation) => {
                violation.nodes.forEach((a11yError) => {
                    const suiteKey = `[Sa11y ${violation.id}]: ${suiteName}`;
                    if (!consolidatedErrors.has(suiteKey)) consolidatedErrors.set(suiteKey, []);
                    consolidatedErrors.get(suiteKey)?.push({
                        ...testResult,
                        fullName: `${a11yError.target[0]}`, // CSS Selector
                        failureMessages: [
                            `Accessibility issue found: ${violation.help}
 Help: ${violation.helpUrl.split('?')[0]}
 CSS Selectors: "${a11yError.target.join('; ')}"
 Tests: "${testResult.fullName}"
 More info: https://sfdc.co/a11y-jest`,
                        ],
                    });
                });
            });
            // Don't report the failure twice
            testResult.status = 'disabled';
            // Suites with only a11y errors should be marked as passed
            testSuite.numFailingTests -= 1;
            // TODO (fix): Remove sa11y msg from test suite message
            //  ANSI codes and test names in suite message makes it difficult
            // testResult.failureMessages = [];
            // testResult.failureDetails = [];
            testSuite.failureMessage = '';
        }
    });
}

/**
 * Custom results processor for a11y results. Only affects JSON results file output.
 * To be used with jest cli options --json --outputFile
 *  e.g. jest --json --outputFile jestResults.json --testResultsProcessor node_modules/@sa11y/jest/dist/resultsProcessor.js
 * Ref: https://jestjs.io/docs/configuration#testresultsprocessor-string
 *  - Mapping of AggregatedResult to JSON format to https://github.com/facebook/jest/blob/master/packages/jest-test-result/src/formatTestResults.ts
 */
export function resultsProcessor(results: AggregatedResult): AggregatedResult {
    // const sa11yResults = [] as TestResult[];

    results.testResults
        .filter((testSuite) => testSuite.numFailingTests > 0)
        .forEach((testSuite) => {
            // Copy test suite meta-data and clear test results
            // const sa11yTestSuite = { ...testSuite, testResults: [] };

            testSuite.testResults
                .filter((testResult) => testResult.status === 'failed')
                .forEach((testResult) => addTestResult(testResult, testSuite));

            // if (sa11yTestSuite.testResults.length > 0) {
            //     sa11yResults.push(sa11yTestSuite);
            // }
        });

    // results.testResults.push(...sa11yResults);

    consolidatedErrors.forEach((testResults, suiteKey) => {
        const sa11ySuite = buildFailureTestResult(suiteKey, new Error() as SerializableError);
        sa11ySuite.testResults = testResults;
        results.testResults.push(sa11ySuite);
    });

    return results;
}

// The processor must be a node module that exports a function
module.exports = resultsProcessor;
