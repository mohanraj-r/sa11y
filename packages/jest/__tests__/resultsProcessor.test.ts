/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { aggregatedTestResults, domWithA11yIssues } from '@sa11y/test-utils';
import resultsProcessor from '../src/resultsProcessor';
import { automaticCheck } from '../src/automatic';
import { addResult, createEmptyTestResult, makeEmptyAggregatedTestResult } from '@jest/test-result';
import { AssertionResult } from '@jest/test-result/build/types';
import { A11yError } from '@sa11y/format';

// Note: To re-generate test data when required (e.g. changes in A11yError)
//  enable the following tests and the code to output file in resultsProcessor.ts
/* eslint-disable jest/no-commented-out-tests,jest/expect-expect */
// eslint-disable-next-line jest/no-disabled-tests
describe.skip('Prepare test data (when A11yError changes)', () => {
    it('with a11y failures', async () => {
        document.body.innerHTML = domWithA11yIssues;
        await automaticCheck({ cleanupAfterEach: false });
    });

    it('with duplicate a11y failures to test consolidation', async () => {
        document.body.innerHTML = domWithA11yIssues;
        await automaticCheck({ cleanupAfterEach: false });
    });

    it('with non-a11y failures', () => {
        expect(1).toBe(2);
    });
});
/* eslint-enable jest/no-commented-out-tests,jest/expect-expect */

describe('Results Processor', () => {
    it.skip('should prepare test data', () => {
        // TODO (refactor): Generate test results using helpers instead of hard-coded result file.
        //  The helpers seem to have functions to generate only empty test results.
        //  Not able to find one which can take an Error() and convert it into a test result (AssertionResult).
        const aggregatedResults = makeEmptyAggregatedTestResult();
        const testSuite = createEmptyTestResult();
        const testResult = { failureDetails: [new A11yError([])], status: 'failed' } as AssertionResult;
        testSuite.testResults.push(testResult);
        testSuite.numFailingTests += 1;
        addResult(aggregatedResults, testSuite);
        expect(resultsProcessor(aggregatedTestResults)).toMatchSnapshot();
    });

    it('should have valid test data to start with', () => {
        expect(aggregatedTestResults.numFailedTests).toBeGreaterThan(0);
    });

    it('should process test results as expected', () => {
        expect(resultsProcessor(aggregatedTestResults)).toMatchSnapshot();
        // TODO (tests): Add more fine grained, targeted tests
        // expect(resultsProcessor(aggregatedTestResults)).not.toStrictEqual(aggregatedTestResults);
    });
});
