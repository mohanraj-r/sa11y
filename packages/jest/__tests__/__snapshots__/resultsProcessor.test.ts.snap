// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results Processor should process test results as expected 1`] = `
Object {
  "numFailedTestSuites": 1,
  "numFailedTests": 3,
  "numPassedTestSuites": 0,
  "numPassedTests": 0,
  "numPendingTestSuites": 0,
  "numPendingTests": 0,
  "numRuntimeErrorTestSuites": 0,
  "numTodoTests": 0,
  "numTotalTestSuites": 1,
  "numTotalTests": 3,
  "openHandles": Array [],
  "snapshot": Object {
    "added": 0,
    "didUpdate": false,
    "failure": false,
    "filesAdded": 0,
    "filesRemoved": 0,
    "filesRemovedList": Array [],
    "filesUnmatched": 0,
    "filesUpdated": 0,
    "matched": 0,
    "total": 0,
    "unchecked": 0,
    "uncheckedKeysByFile": Array [],
    "unmatched": 0,
    "updated": 0,
  },
  "startTime": 1622770460154,
  "success": false,
  "testResults": Array [
    Object {
      "failureMessage": "[1m[31m  [1m● [22m[1mPrepare test data › with a11y failures[39m[22m

    A11yError: 1 accessibility issues found
     * (link-name) Links must have discernible text: a
    	- Help URL: https://dequeuniversity.com/rules/axe/4.1/link-name
[2m[22m
[2m    [0m [90m 65 |[39m         }[0m[22m
[2m    [0m [90m 66 |[39m         [36mif[39m (violations[33m.[39mlength [33m>[39m [35m0[39m) {[0m[22m
[2m    [0m[31m[1m>[22m[2m[39m[90m 67 |[39m             [36mthrow[39m [36mnew[39m [33mA11yError[39m(violations[33m,[39m opts)[33m;[39m[0m[22m
[2m    [0m [90m    |[39m                   [31m[1m^[22m[2m[39m[0m[22m
[2m    [0m [90m 68 |[39m         }[0m[22m
[2m    [0m [90m 69 |[39m     }[0m[22m
[2m    [0m [90m 70 |[39m[0m[22m
[2m[22m
[2m      [2mat Function.checkAndThrow ([22m[2mpackages/format/src/format.ts[2m:67:19)[22m[2m[22m
[2m      [2mat automaticCheck ([22m[2mpackages/jest/src/automatic.ts[2m:54:19)[22m[2m[22m
[2m      [2mat Object.<anonymous> ([22m[2m[0m[36mpackages/jest/__tests__/resultsProcessor.test.ts[39m[0m[2m:31:9)[22m[2m[22m

[1m[31m  [1m● [22m[1mPrepare test data › with duplicate a11y failures to test consolidation[39m[22m

    A11yError: 1 accessibility issues found
     * (link-name) Links must have discernible text: a
    	- Help URL: https://dequeuniversity.com/rules/axe/4.1/link-name
[2m[22m
[2m    [0m [90m 65 |[39m         }[0m[22m
[2m    [0m [90m 66 |[39m         [36mif[39m (violations[33m.[39mlength [33m>[39m [35m0[39m) {[0m[22m
[2m    [0m[31m[1m>[22m[2m[39m[90m 67 |[39m             [36mthrow[39m [36mnew[39m [33mA11yError[39m(violations[33m,[39m opts)[33m;[39m[0m[22m
[2m    [0m [90m    |[39m                   [31m[1m^[22m[2m[39m[0m[22m
[2m    [0m [90m 68 |[39m         }[0m[22m
[2m    [0m [90m 69 |[39m     }[0m[22m
[2m    [0m [90m 70 |[39m[0m[22m
[2m[22m
[2m      [2mat Function.checkAndThrow ([22m[2mpackages/format/src/format.ts[2m:67:19)[22m[2m[22m
[2m      [2mat automaticCheck ([22m[2mpackages/jest/src/automatic.ts[2m:54:19)[22m[2m[22m
[2m      [2mat Object.<anonymous> ([22m[2m[0m[36mpackages/jest/__tests__/resultsProcessor.test.ts[39m[0m[2m:36:9)[22m[2m[22m

[1m[31m  [1m● [22m[1mPrepare test data › with non-a11y failures[39m[22m

    [2mexpect([22m[31mreceived[39m[2m).[22mtoBe[2m([22m[32mexpected[39m[2m) // Object.is equality[22m

    Expected: [32m2[39m
    Received: [31m1[39m
[2m[22m
[2m    [0m [90m 38 |[39m[0m[22m
[2m    [0m [90m 39 |[39m     it([32m'with non-a11y failures'[39m[33m,[39m () [33m=>[39m {[0m[22m
[2m    [0m[31m[1m>[22m[2m[39m[90m 40 |[39m         expect([35m1[39m)[33m.[39mtoBe([35m2[39m)[33m;[39m[0m[22m
[2m    [0m [90m    |[39m                   [31m[1m^[22m[2m[39m[0m[22m
[2m    [0m [90m 41 |[39m     })[33m;[39m[0m[22m
[2m    [0m [90m 42 |[39m })[33m;[39m[0m[22m
[2m    [0m [90m 43 |[39m [90m/* eslint-enable jest/no-commented-out-tests */[39m[0m[22m
[2m[22m
[2m      [2mat Object.<anonymous> ([22m[2m[0m[36mpackages/jest/__tests__/resultsProcessor.test.ts[39m[0m[2m:40:19)[22m[2m[22m
",
      "leaks": false,
      "numFailingTests": 1,
      "numPassingTests": 0,
      "numPendingTests": 0,
      "numTodoTests": 0,
      "openHandles": Array [],
      "perfStats": Object {
        "end": 1622770462152,
        "runtime": 1120,
        "slow": false,
        "start": 1622770461032,
      },
      "skipped": false,
      "snapshot": Object {
        "added": 0,
        "fileDeleted": false,
        "matched": 0,
        "unchecked": 0,
        "uncheckedKeys": Array [],
        "unmatched": 0,
        "updated": 0,
      },
      "testFilePath": "/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/packages/jest/__tests__/resultsProcessor.test.ts",
      "testResults": Array [
        Object {
          "ancestorTitles": Array [
            "Prepare test data",
          ],
          "duration": 148,
          "failureDetails": Array [
            Object {
              "actual": "",
              "error": Object {
                "a11yResults": Array [
                  Object {
                    "description": "Links must have discernible text",
                    "helpUrl": "https://dequeuniversity.com/rules/axe/4.1/link-name",
                    "html": "<a href=\\"#\\"></a>",
                    "id": "link-name",
                    "selectors": "a",
                    "wcag": Object {
                      "successCriteria": "2.4.4",
                      "tags": Array [
                        "cat.name-role-value",
                        "wcag2a",
                        "wcag412",
                        "wcag244",
                        "section508",
                        "section508.22.a",
                        "ACT",
                      ],
                      "wcagLevel": "A",
                      "wcagVersion": "2.0",
                    },
                  },
                ],
                "name": "A11yError",
                "violations": Array [
                  Object {
                    "description": "Ensures links have discernible text",
                    "help": "Links must have discernible text",
                    "helpUrl": "https://dequeuniversity.com/rules/axe/4.1/link-name?application=axeAPI",
                    "id": "link-name",
                    "impact": "serious",
                    "nodes": Array [
                      Object {
                        "all": Array [],
                        "any": Array [
                          Object {
                            "data": null,
                            "id": "has-visible-text",
                            "impact": "minor",
                            "message": "Element does not have text that is visible to screen readers",
                            "relatedNodes": Array [],
                          },
                          Object {
                            "data": null,
                            "id": "aria-label",
                            "impact": "serious",
                            "message": "aria-label attribute does not exist or is empty",
                            "relatedNodes": Array [],
                          },
                          Object {
                            "data": null,
                            "id": "aria-labelledby",
                            "impact": "serious",
                            "message": "aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty",
                            "relatedNodes": Array [],
                          },
                          Object {
                            "data": Object {
                              "messageKey": "noAttr",
                            },
                            "id": "non-empty-title",
                            "impact": "serious",
                            "message": "Element has no title attribute",
                            "relatedNodes": Array [],
                          },
                        ],
                        "failureSummary": "Fix all of the following:
  Element is in tab order and does not have accessible text

Fix any of the following:
  Element does not have text that is visible to screen readers
  aria-label attribute does not exist or is empty
  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
  Element has no title attribute",
                        "html": "<a href=\\"#\\"></a>",
                        "impact": "serious",
                        "none": Array [
                          Object {
                            "data": null,
                            "id": "focusable-no-name",
                            "impact": "serious",
                            "message": "Element is in tab order and does not have accessible text",
                            "relatedNodes": Array [],
                          },
                        ],
                        "target": Array [
                          "a",
                        ],
                      },
                    ],
                    "tags": Array [
                      "cat.name-role-value",
                      "wcag2a",
                      "wcag412",
                      "wcag244",
                      "section508",
                      "section508.22.a",
                      "ACT",
                    ],
                  },
                ],
              },
              "expected": "",
              "matcherName": "",
              "message": "A11yError: 1 accessibility issues found
 * (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.1/link-name",
              "passed": false,
              "stack": "A11yError: 1 accessibility issues found
 * (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.1/link-name
    at Function.checkAndThrow (/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/packages/format/src/format.ts:67:19)
    at automaticCheck (/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/packages/jest/src/automatic.ts:54:19)
    at Object.<anonymous> (/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/packages/jest/__tests__/resultsProcessor.test.ts:31:9)",
            },
          ],
          "failureMessages": Array [
            "A11yError: 1 accessibility issues found
 * (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.1/link-name
    at Function.checkAndThrow (/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/packages/format/src/format.ts:67:19)
    at automaticCheck (/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/packages/jest/src/automatic.ts:54:19)
    at Object.<anonymous> (/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/packages/jest/__tests__/resultsProcessor.test.ts:31:9)",
          ],
          "fullName": "Prepare test data with a11y failures",
          "location": null,
          "numPassingAsserts": 0,
          "status": "disabled",
          "title": "with a11y failures",
        },
        Object {
          "ancestorTitles": Array [
            "Prepare test data",
          ],
          "duration": 43,
          "failureDetails": Array [
            Object {
              "actual": "",
              "error": Object {
                "a11yResults": Array [
                  Object {
                    "description": "Links must have discernible text",
                    "helpUrl": "https://dequeuniversity.com/rules/axe/4.1/link-name",
                    "html": "<a href=\\"#\\"></a>",
                    "id": "link-name",
                    "selectors": "a",
                    "wcag": Object {
                      "successCriteria": "2.4.4",
                      "tags": Array [
                        "cat.name-role-value",
                        "wcag2a",
                        "wcag412",
                        "wcag244",
                        "section508",
                        "section508.22.a",
                        "ACT",
                      ],
                      "wcagLevel": "A",
                      "wcagVersion": "2.0",
                    },
                  },
                ],
                "name": "A11yError",
                "violations": Array [
                  Object {
                    "description": "Ensures links have discernible text",
                    "help": "Links must have discernible text",
                    "helpUrl": "https://dequeuniversity.com/rules/axe/4.1/link-name?application=axeAPI",
                    "id": "link-name",
                    "impact": "serious",
                    "nodes": Array [
                      Object {
                        "all": Array [],
                        "any": Array [
                          Object {
                            "data": null,
                            "id": "has-visible-text",
                            "impact": "minor",
                            "message": "Element does not have text that is visible to screen readers",
                            "relatedNodes": Array [],
                          },
                          Object {
                            "data": null,
                            "id": "aria-label",
                            "impact": "serious",
                            "message": "aria-label attribute does not exist or is empty",
                            "relatedNodes": Array [],
                          },
                          Object {
                            "data": null,
                            "id": "aria-labelledby",
                            "impact": "serious",
                            "message": "aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty",
                            "relatedNodes": Array [],
                          },
                          Object {
                            "data": Object {
                              "messageKey": "noAttr",
                            },
                            "id": "non-empty-title",
                            "impact": "serious",
                            "message": "Element has no title attribute",
                            "relatedNodes": Array [],
                          },
                        ],
                        "failureSummary": "Fix all of the following:
  Element is in tab order and does not have accessible text

Fix any of the following:
  Element does not have text that is visible to screen readers
  aria-label attribute does not exist or is empty
  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
  Element has no title attribute",
                        "html": "<a href=\\"#\\"></a>",
                        "impact": "serious",
                        "none": Array [
                          Object {
                            "data": null,
                            "id": "focusable-no-name",
                            "impact": "serious",
                            "message": "Element is in tab order and does not have accessible text",
                            "relatedNodes": Array [],
                          },
                        ],
                        "target": Array [
                          "a",
                        ],
                      },
                    ],
                    "tags": Array [
                      "cat.name-role-value",
                      "wcag2a",
                      "wcag412",
                      "wcag244",
                      "section508",
                      "section508.22.a",
                      "ACT",
                    ],
                  },
                ],
              },
              "expected": "",
              "matcherName": "",
              "message": "A11yError: 1 accessibility issues found
 * (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.1/link-name",
              "passed": false,
              "stack": "A11yError: 1 accessibility issues found
 * (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.1/link-name
    at Function.checkAndThrow (/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/packages/format/src/format.ts:67:19)
    at automaticCheck (/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/packages/jest/src/automatic.ts:54:19)
    at Object.<anonymous> (/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/packages/jest/__tests__/resultsProcessor.test.ts:36:9)",
            },
          ],
          "failureMessages": Array [
            "A11yError: 1 accessibility issues found
 * (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.1/link-name
    at Function.checkAndThrow (/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/packages/format/src/format.ts:67:19)
    at automaticCheck (/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/packages/jest/src/automatic.ts:54:19)
    at Object.<anonymous> (/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/packages/jest/__tests__/resultsProcessor.test.ts:36:9)",
          ],
          "fullName": "Prepare test data with duplicate a11y failures to test consolidation",
          "location": null,
          "numPassingAsserts": 0,
          "status": "disabled",
          "title": "with duplicate a11y failures to test consolidation",
        },
        Object {
          "ancestorTitles": Array [
            "Prepare test data",
          ],
          "duration": 5,
          "failureDetails": Array [
            Object {
              "actual": "",
              "error": Object {
                "matcherResult": Object {
                  "actual": 1,
                  "expected": 2,
                  "name": "toBe",
                  "pass": false,
                },
              },
              "expected": "",
              "matcherName": "",
              "message": "Error: [2mexpect([22m[31mreceived[39m[2m).[22mtoBe[2m([22m[32mexpected[39m[2m) // Object.is equality[22m

Expected: [32m2[39m
Received: [31m1[39m",
              "passed": false,
              "stack": "Error: [2mexpect([22m[31mreceived[39m[2m).[22mtoBe[2m([22m[32mexpected[39m[2m) // Object.is equality[22m

Expected: [32m2[39m
Received: [31m1[39m
    at Object.<anonymous> (/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/packages/jest/__tests__/resultsProcessor.test.ts:40:19)
    at Object.asyncJestTest (/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:106:37)
    at /Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/node_modules/jest-jasmine2/build/queueRunner.js:45:12
    at new Promise (<anonymous>)
    at mapper (/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/node_modules/jest-jasmine2/build/queueRunner.js:28:19)
    at /Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/node_modules/jest-jasmine2/build/queueRunner.js:75:41",
            },
          ],
          "failureMessages": Array [
            "Error: [2mexpect([22m[31mreceived[39m[2m).[22mtoBe[2m([22m[32mexpected[39m[2m) // Object.is equality[22m

Expected: [32m2[39m
Received: [31m1[39m
    at Object.<anonymous> (/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/packages/jest/__tests__/resultsProcessor.test.ts:40:19)
    at Object.asyncJestTest (/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:106:37)
    at /Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/node_modules/jest-jasmine2/build/queueRunner.js:45:12
    at new Promise (<anonymous>)
    at mapper (/Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/node_modules/jest-jasmine2/build/queueRunner.js:28:19)
    at /Users/mrajamanickam/src/a11y/sa11y/github.com/sa11y/node_modules/jest-jasmine2/build/queueRunner.js:75:41",
          ],
          "fullName": "Prepare test data with non-a11y failures",
          "location": null,
          "numPassingAsserts": 0,
          "status": "failed",
          "title": "with non-a11y failures",
        },
      ],
    },
    Object {
      "console": undefined,
      "displayName": undefined,
      "failureMessage": null,
      "leaks": false,
      "numFailingTests": 0,
      "numPassingTests": 0,
      "numPendingTests": 0,
      "numTodoTests": 0,
      "openHandles": Array [],
      "perfStats": Object {
        "end": 0,
        "runtime": 0,
        "slow": false,
        "start": 0,
      },
      "skipped": false,
      "snapshot": Object {
        "added": 0,
        "fileDeleted": false,
        "matched": 0,
        "unchecked": 0,
        "uncheckedKeys": Array [],
        "unmatched": 0,
        "updated": 0,
      },
      "sourceMaps": Object {},
      "testExecError": [Error],
      "testFilePath": "[Sa11y [object Object] link-name]: resultsProcessor.test.ts",
      "testResults": Array [
        Object {
          "ancestorTitles": Array [
            "Prepare test data",
            "Prepare test data with a11y failures",
          ],
          "duration": 148,
          "failureDetails": Array [],
          "failureMessages": Array [
            "accessibility issues found: Links must have discernible text
 CSS Selectors: a
 HTML element: <a href=\\"#\\"></a>
 Help: https://dequeuniversity.com/rules/axe/4.1/link-name
 Tests: \\"Prepare test data with a11y failures\\"",
          ],
          "fullName": "Links must have discernible text",
          "location": null,
          "numPassingAsserts": 0,
          "status": "disabled",
          "title": "with a11y failures",
        },
        Object {
          "ancestorTitles": Array [
            "Prepare test data",
            "Prepare test data with duplicate a11y failures to test consolidation",
          ],
          "duration": 43,
          "failureDetails": Array [],
          "failureMessages": Array [
            "accessibility issues found: Links must have discernible text
 CSS Selectors: a
 HTML element: <a href=\\"#\\"></a>
 Help: https://dequeuniversity.com/rules/axe/4.1/link-name
 Tests: \\"Prepare test data with duplicate a11y failures to test consolidation\\"",
          ],
          "fullName": "Links must have discernible text",
          "location": null,
          "numPassingAsserts": 0,
          "status": "disabled",
          "title": "with duplicate a11y failures to test consolidation",
        },
      ],
    },
  ],
  "wasInterrupted": false,
}
`;
